src/SystemManager.cc

```diff
@@ -460,58 +452,41 @@ void SystemManager::ProcessRemovedEntities(
     return;
   }
 
-  RemoveFromVectorIf(this->systemsReset,
-    [&](const SystemIfaceWithParent<ISystemReset>& system) {
-      return _ecm.IsMarkedForRemoval(system.parent);
-    });
-  RemoveFromVectorIf(this->systemsPreupdate,
-    [&](const SystemIfaceWithParent<ISystemPreUpdate>& system) {
-      return _ecm.IsMarkedForRemoval(system.parent);
-    });
-  RemoveFromVectorIf(this->systemsUpdate,
-    [&](const SystemIfaceWithParent<ISystemUpdate>& system) {
-      return _ecm.IsMarkedForRemoval(system.parent);
-    });
+  std::vector<SystemInternal> markedForRemoval;
+  for (const auto &system : this->systems)
+  {
+    if (_ecm.IsMarkedForRemoval(system.parentEntity))
+    {
+      markedForRemoval.push_back(system);
+    }
+  }
 
-  std::unordered_set<ISystemPostUpdate *> markedForRemoval;
-  for (std::size_t i = 0; i < this->systemsPostupdate.size(); i++)
+  for (const auto &system : markedForRemoval)
   {
-    if(_ecm.IsMarkedForRemoval(postUpdateParents[i]))
+    removeFromVector(this->systemsReset, system.reset);
+    removeFromVector(this->systemsPreupdate, system.preupdate);
+    removeFromVector(this->systemsUpdate, system.update);
+
+    if (system.postupdate)
     {
       // If system with a PostUpdate is marked for removal
       // mark all worker threads for removal.
       _needsCleanUp = true;
-      markedForRemoval.insert(this->systemsPostupdate[i]);
     }
-  }
 
-  RemoveFromVectorIf(this->systemsPostupdate,
-    [&](const auto system) {
-      if (markedForRemoval.count(system)) {
-        return true;
-      }
-      return false;
-    });
+    removeFromVector(this->systemsPostupdate, system.postupdate);
```

Comment:
I tested this. It works just fine. The only thing I'd say is that we do some extra work by iterating over each system (`n`) by the number of entities removed (`m`)  (time complexity `O(mn)`), rather than just each system (which is what the original PR does) (time complexity `O(n)`). Its fine if we have only a few entities removed at a time. If you'd like I can fix that by building `unordered_sets` of `ISystemPreupdate*` or we can just leave it as is.

Reply:
OK Opened a PR to this PR implementing the changes:
#2475